### ISSUE TITLE
IdentityServer does not appear to use CORS policy service

### ISSUE THREAD

**Issue by ChrisSwinchatt** [IdentityServer/IdentityServer4#3075]
### Issue / Steps to reproduce the problem

I have created an `ICorsPolicyService` that always returns true because we prefer to handle CORS in the API gateway and have the downstream services ignore CORS. However, it seems that IS4 does not use my CORS policy.

I register my policy service like this:

```csharp
services.AddTransient<ICorsPolicyService, MyCorsPolicyService>(); // Added this for testing, makes no apparent difference

services
    .AddIdentityServer(o =>
    {
        o.PublicOrigin = Options.Token.Issuer;
    })
    .AddProfileService<MyProfileService>()
    .AddCorsPolicyService<MyCorsPolicyService>()
;
```

And the policy looks like this:

```csharp
public Task<bool> IsOriginAllowedAsync(string origin)
{
    _logger.LogInformation($"Allowing CORS origin {origin}");
    return Task.FromResult(true);
}
```

Which, AFAICT, should allow everything. The ASP.Net Core CORS configuration is wildcarded so it should be allowing any origin as well. However, the log doesn't have my "Allowing CORS origin" message anywhere which implies it's never being run. Strangely, sometimes an origin will be allowed, and sometimes it won't.

### Relevant parts of the log file

```
[09:55:56 DBG] CORS request made for path: /connect/token from origin: https://my.website.com
[09:55:56 DBG] Origin https://my.website.com is allowed: True
[09:55:56 DBG] CorsPolicyService allowed origin: https://my.website.com
...

[10:14:00 DBG] CORS request made for path: /connect/token from origin: http://localhost:8080
[10:14:00 DBG] Origin http://localhost:8080 is allowed: False
[10:14:00 WRN] CorsPolicyService did not allow origin: http://localhost:8080
```

I'm not sure how IS4 is deciding whether to allow these origins but it doesn't seem to use my policy service in any case.

**Comment by ChrisSwinchatt** [comment]
Also, interestingly, the CORS requests do seem to be allowed anyway - the browser doesn't report a problem.

**Comment by othman-fayed** [comment]
CORS in IDS4 allows the registered client's urls by default as per Client.AllowedCorsOrigins

**Comment by ChrisSwinchatt** [comment]
Thank you for that page, but I'm not sure it answers my question of **why my CORS policy service is not being called**. The documentation states that IS4 allows the application to control the CORS policy completely ([source](http://docs.identityserver.io/en/latest/topics/cors.html?highlight=IdentityServer%20allows%20the%20hosting%20application%20to%20implement%20the%20ICorsPolicyService%20to%20completely%20control%20the%20CORS%20policy.#client-based-cors-configuration)) but I am unable to get this work.

**Comment by brockallen** [comment]
Hmm, it should be called. Are you testing with a JS client? Do you see the Origin in the request headers and do you see preflight requests from the browser?

**Comment by ChrisSwinchatt** [comment]
Testing with Postman, if I include the origin header, regardless of the value, I get:

```
[08:30:42 DBG] CORS request made for path: /connect/token from origin: https://product.company.com
[08:30:42 DBG] Origin https://product.company.com is allowed: False
```

It doesn't seem to matter what the origin is, even though my allowed origins in ASP.Net Core's CORS configuration is set to allow all origins, headers, methods, etc.

**Comment by brockallen** [comment]
> Testing with Postman 

Don't, please. Use a real JS client in a browser.

**Comment by brockallen** [comment]
Any updates on this from your side?

**Comment by ChrisSwinchatt** [comment]
Everything is working from the frontend but we are still seeing those log messages. We transferred responsibility for CORS into a gateway and every other service just accepts any CORS request that gets through. This works fine despite the log messages, so we have just moved on.

I'm sorry I wasn't able to repro this with a browser client, but we couldn't allocate time to it as it would have blocked our front-end team.
